framework/auth/views.py

```diff
@@ -237,26 +238,28 @@ def login_and_register_handler(auth, login=True, campaign=None, next_url=None, l
                 else:
                     data['next_url'] = cas.get_login_url(next_url, campaign='institution')
             # for non-institution campaigns
-            elif auth.logged_in:
-                # if user is already logged in, go to the campaign landing page
-                data['status_code'] = http.FOUND
-                data['next_url'] = campaigns.campaign_url_for(campaign)
             else:
-                # if user is logged out, go to the osf register page with campaign context
-                if login:
-                    # `GET /login?campaign=...`
-                    data['next_url'] = web_url_for('auth_register', campaign=campaign)
+                destination = next_url if next_url else campaigns.campaign_url_for(campaign)
```

Comment:
This `destination` is used for `next=` query parameter in `/login?` and `/register?` urls. Valid `destination` overwrites default `next_url`. For example, instead of landing user on campaign register page, `next=` can land user on specific preprint page.